src/nf/nf_optimizers.f90

```diff
@@ -123,4 +144,44 @@ pure subroutine minimize_rmsprop(self, param, gradient)
 
   end subroutine minimize_rmsprop
 
+
+  impure elemental subroutine init_adam(self, num_params)
+    class(adam), intent(inout) :: self
+    integer, intent(in) :: num_params
+    if (.not. allocated(self % m)) then
```

Comment:
Could `v` be allocated and not `m`?
I suggest to check the status of both `m` and `v`.

Reply:
That would be an error state, see https://github.com/modern-fortran/neural-fortran/pull/150#discussion_r1266806684. `v` and `m` will always have the same allocation status.

@Spnetic-5 let's make `m` and `v` `private` so that the user can't fiddle with them directly.

I'm also thinking whether we should make the `init` method `private` and invoke it from `minimize`. This decreases the number of places in the code at the network level that the optimizer instance is modified.